transmart-core-db/grails-app/services/org/transmartproject/db/clinical/CrossTableService.groovy

```diff
@@ -2,66 +2,108 @@ package org.transmartproject.db.clinical
 
 import groovy.transform.CompileStatic
 import org.springframework.beans.factory.annotation.Autowired
-import org.transmartproject.core.exceptions.AccessDeniedException
+import org.transmartproject.core.multidimquery.AggregateDataResource
 import org.transmartproject.core.multidimquery.CrossTable
 import org.transmartproject.core.multidimquery.CrossTableResource
 import org.transmartproject.core.multidimquery.PatientSetResource
 import org.transmartproject.core.multidimquery.query.AndConstraint
 import org.transmartproject.core.multidimquery.query.Constraint
-import org.transmartproject.core.multidimquery.query.MultipleSubSelectionsConstraint
-import org.transmartproject.core.multidimquery.query.Operator
 import org.transmartproject.core.multidimquery.query.PatientSetConstraint
-import org.transmartproject.core.multidimquery.query.SubSelectionConstraint
+import org.transmartproject.core.querytool.QueryResult
 import org.transmartproject.core.users.PatientDataAccessLevel
 import org.transmartproject.core.users.User
 
+import java.util.function.BiFunction
+import java.util.stream.Collectors
+
 @CompileStatic
 class CrossTableService extends AbstractDataResourceService implements CrossTableResource {
 
     @Autowired
     PatientSetResource patientSetResource
 
+    @Autowired
+    AggregateDataOptimisationsService aggregateDataOptimisationsService
+
+    @Autowired
+    AggregateDataResource aggregateDataResource
```

Comment:
Maybe here also the service can be used directly instead of core-api interface or  we could add an interface for AggregateDataOptimisationsService?

Reply:
I think the talking via the interface is fine. I'm reluctant to create an interface for `AggregateDataOptimisationsService`. Ideally, it should be another implementation of the same  `AggregateDataResource` interface.